### PR TITLE
test: Add frontend tests for #210

### PR DESCRIPTION
**Pull request by franf91**
# Quick Overview
This pr does not close the issue #210. This pr is for configuring cypress, adding some e2e tests, and updating the cicd to run these tests. More tests need to be added in the next iteration.

In this iteration, 4 specs were created: Registration, Login, Admin, and Setting, comprising a total of 12 tests.

- Registration Spec: Tests both error and successful registration flows. Error scenarios include cases with missing inputs, mismatched passwords, and duplicate emails.
- Login Spec: Focuses on error and successful login flows. Error cases include missing inputs and invalid credentials, such as incorrect passwords or non-existent emails.
- Admin Spec: Covers error and successful admin creation flows, testing scenarios like missing inputs, mismatched passwords, and duplicate emails. It also includes tests for approving and rejecting client-submitted projects.
- Setting Spec: Validates error and successful password change flows on the settings page. Error scenarios include missing inputs, mismatched new and confirm passwords, and incorrect current passwords.

Below explains what is cypress, the installation process, additional configuration, some key cypress concepts, explaining 1 of the 12  tests ( No Error Registration Flow For Freelancer Option) and changes made to CICD.

# What is Cypress

Cypress is a front end testing tool build for the modern web. Cypress provides solutions for End-to-end testing, component testing, accessibility testing, ui coverage, and more.

# Installing Cypress for  the first time
```sh
npm install cypress --save-dev
```

# Open Cypress for the first time
```sh
npx cypress open
```
# Creating first spec

1.  Select E2E or Component Testing
Screen Shot 2025-01-09 at 2 32 52 PM

2.  Add Configuration Files
Screen Shot 2025-01-09 at 2 34 34 PM

3. Select A Browser For Testing
Screen Shot 2025-01-09 at 2 36 34 PM

4. Create And Run First Spec
Screen Shot 2025-01-09 at 2 38 41 PM

Screen Shot 2025-01-09 at 2 39 55 PM

Screen Shot 2025-01-09 at 2 40 55 PM

Screen Shot 2025-01-09 at 2 42 33 PM

# Additional Configuration

1.  Make Base Url localhost:3000 by adding following to the cypress.config.ts file under e2e

```sh
baseUrl:'http://localhost:3000'
```
2. Change open command by updating the package.json file under scripts

```sh
npm run cy:open
```
3. Add an additional command than runs all tests in terminal instead of browser by updating the package.json file under scripts
```sh
npm run e2e:chrome
```

# What is a spec?

- A spec in Cypress refers to a test file, typically written in JavaScript or TypeScript, where you define your end-to-end tests. Each spec file organizes and groups related tests for a particular feature or functionality of your application, making it easier to maintain and execute targeted test cases.

#  What is a Describe block?

- The describe block in Cypress is used to group related tests into a suite, providing a logical structure to your test files. It allows you to name the group and acts as a container for organizing multiple it blocks that test specific scenarios within the same feature or context.

# What is an IT block?

- An it block in Cypress represents an individual test case or scenario within a test suite. It defines the specific behavior or functionality you want to test, using Cypress commands to interact with and assert against your application.

#  Selectors

- Cypress recommends using selectors like data-cy, data-test, and data-testid to target elements in your code for testing. These attributes are specifically designated for testing purposes, ensuring that your tests remain stable and unaffected by changes to classes or IDs that are primarily used for styling or functionality.

# The cy object

- The cy object is the core interface in Cypress, used to interact with elements in your application. It provides various methods like get(), contains(), type(), visit(), click(), and  intercept()

# The get() method

-  This method is used to target a given selector. For example the following selects the data-test selector with the name "firstName-input":
```sh
cy.get('[data-test="firstName-input"]')
```
# The contains() method
- This method is used to verify whether a specific element or text is present within the selected code. For example, the following contains()  verifies if the error message "A user with this email already exists" is displayed:

```sh
cy.get('[data-test="error-message"]').contains(/A user with this email already exists./i)
```
# The type() method

- This method enables Cypress to simulate typing into an appropriate HTML element, such as an input field. For example, the following code types "John" into the first name input field:

```sh
cy.get('[data-test="firstName-input"]').type("John")
```
# The visit() method

- This method enables Cypress to simulate the navigation to a url. For example, the following visits the login page:

```sh
cy.visit('/auth/login')
```
# The click() method
- This method enables Cypress to simulate a click event. For example, the following clicks on the create admin button

```sh
cy.get('[data-test="create-admin-btn"]').click()
```
# The intercept() method
- This method can be used to mock api responses from the backend. It's meant to intercept the request before reaching backend and return a mock response.For example, the following mocks an error response for the POST request to the endpoint "/api/users/". The appropriate status code and return message is added. The .as('registerUserError') is an alias that can be used to call this intercept at any point in the cypress code:

```code
cy.intercept('POST', '/api/users', {
  statusCode: 400,
  body: {
    message: 'A user with this email already exists.',
  },
}).as('registerUserError');
```
# The test : No Error Registration Flow For Freelancer Option

- The following test demonstrates the successful flow for signing up as a freelancer. It begins by navigating to the /auth/register page, selecting the freelancer option, and clicking the button to proceed to the form. The test then fills out the registration form, submits it, and mocks the API success response. Finally, it verifies that the success message is displayed.

``` code
  it("No Error Registration Flow For Freelancer Option",()=>{
    
    // Mock api success response
    cy.intercept('POST', '/api/users', {
      statusCode: 200,
      body: {
        message: 'Account created successfully',
      },
    }).as('registerUserSuccess');
 
 
    // Visit the registration page
    cy.visit('/auth/register')
 
 
    // Select the "Freelancer" option
    cy.get('[data-test="freelancer-role"]').click()
 
 
    // Click the submit button to proceed
    cy.get('[data-test="select-role"]').click()
 
 
    // Fill out the registration form
    cy.get('[data-test="firstName-input"]').type("John");
    cy.get('[data-test="lastName-input"]').type("Doe");
    cy.get('[data-test="email-input"]').type("freelancer@freelancer.com");
    cy.get('[data-test="password-input"]').type("Test1234!");
    cy.get('[data-test="passwordConfirm-input"]').type("Test1234!");
 
 
    // Submit the form
    cy.get('[data-test="register-submit"]').click()
 
 
    // Wait for the mocked API request
    cy.wait('@registerUserSuccess');
 
 
    // Verify the success message is displayed
    cy.get('[data-test="success-message"]').contains(/Account Created/i)
 
 
  })
 })
```
# Updating The CICD main.yml file to run cypress tests

- Two steps are added to the file. The step 2 is to start the front end at localhost:3000 as cypress requires it. The step 3 is to run the cypress e2e tests.

``` code 
   # Step 2: Start the Frontend Server
      - name: Start Frontend Server
        run: |
          cd quolance-ui
          npm run dev &
        env:
          PORT: 3000
      
      # Step 3: Run Cypress Tests
      - name: Run Cypress Tests
        run: |
          cd quolance-ui
          npm run e2e:chrome
```

### CODE REVIEW

File: quolance-ui/cypress.config.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { defineConfig } from "cypress";
+
+export default defineConfig({
+  e2e: {
+    baseUrl:'http://localhost:3000',
```
Comment (KA-devl) [inline quolance-ui/cypress.config.ts:5]: could this potentially cause a clash with our next app being on local host 3000?

**Review by KA-devl — APPROVED** [review]

**Review by abdelh17 — APPROVED** [review]
